### ISSUE TITLE
Support for links like `[this]`?

### ISSUE THREAD

**Issue by rudymatela** [trentm/python-markdown2#597]
Consider the following markdown file `links.md`:

```md
### Reference-style links

This is a link to the [Python](https://python.org) website.

This is a link to the [Python] website.

This is a link to the [Python][] website.

[Python]: https://python.org
```

I can compile it with markdown2 as follows:

```md
markdown2 links.md > links.html
```

When I open the HTML, this is what I get:

![links](https://github.com/user-attachments/assets/06044ade-7dd3-4183-94d6-6470cb0b070e)

1. The first link to python.org renders as "[Python](https://python.org)";
2. The second link to python.org literally renders as `[Python]`;
3. The third link renders as "[Python][]"

[Python]: https://python.org

Is writing links as just `[Python]` supported? Or are only links with `[Python][]` supported?

Some other markdown engines (e.g.: GFM) support `[Python]`.  Could this be added to `markdown2` as an extra to be activated with `-x`?

I actually initially thought that `[Python]` was standard markdown.  Upon looking at the [syntax specification][] it seems only `[Python][]` would be standard.  It nevertheless seems to be that the former is quite common, so I suggest adding it as an extension.  (If you think this is a good idea of course.)

[syntax specification]: https://daringfireball.net/projects/markdown/syntax#link

### Debug info

```bash
$ python --version
Python 3.12.4

$ markdown2 --version
markdown2 2.5.0
```

I also tried with:

```md
markdown2 -x footnotes links.md > links.html
```

**Comment by nicholasserra** [comment]
Interesting, I wouldn't be opposed to it. Assuming there's a good name for the extra lol.

**Comment by rudymatela** [comment]
@nicholasserra GFM calls `[this]` a [shortcut reference link](https://github.github.com/gfm/#shortcut-reference-link).  I know that some other engines support this, but I don't know what they call it.  "Shortcut reference link" seems appropriate enough. 

`[this]: https://example.com`

In terms of the "extra" name on `--help`, you could have it called:

1. `shortcut-reference-link`
2. `short-links`
3. `link-shortcuts`
4. `link-short`
5. `link-shortref`

Of the above, I personally would prefer 4 (`link-short`) as it would appear beside the related `link-patterns` in `markdown2 --help`.  It is also brief.  One suggestion for the help message would be _"`link-short`: allow shortcut reference links, not followed by `[]` or a link label"_

```
$ markdown2 --help
usage: markdown2 [PATHS...]

...
...
...
* html-classes: ...
* link-patterns: Auto-link given regex patterns in text (e.g. bug number
  references, revision number references).
* link-short: allow shortcut reference links, not followed by [] or a link label.
* markdown-in-html: ...
...
...
...
```

**Comment by nicholasserra** [comment]
It seems like we support the full reference links right now, without an extra. Wondering if we should just support the short without an extra too. This is just from a quick glance at the _do_links method. I'll have to do some testing.

**Comment by Crozzers** [comment]
I would agree with that approach.

I took a look at implementing this as a new style extra class and it's tricky to nail down. I settled on detecting snippets like `[Python]` and just adding the `[]` after them so that `_do_links` can process it normally. However, this didn't work particularly well and there were loads of edge cases.
Unless we spun the link processing out into a seperate link processing class (like the `ItalicAndBoldProcessor`) I don't think this approach is feasible.

I think making this part of `_do_links` is probably the most straightforward approach, and I don't mind if we gate it behind an extra or if we make it the default behaviour
